### PR TITLE
Add data.md

### PR DESCRIPTION
**Pull request by xiemoyuan**
Add data.md for paddlenlp.data APIs.

### CODE REVIEW

File: PaddleNLP/docs/data.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+ [8, 9, 0, 0]]
+'''
+```
+
+### `paddlenlp.data.Tuple`
+
+```python
+from paddle.data import Stack, Pad, Tuple
+batchify_fn = Tuple(Pad(axis=0, pad_val=0), Stack())
+```
```
Comment (ZeyuChen) [inline PaddleNLP/docs/data.md:215]: 这个有例子吗？看不太出Tuple的作用

File: PaddleNLP/docs/data.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+| `paddlenlp.data.Pad`            | 堆叠N个输入数据来构建一个batch，每个输入数据将会被padding到N个输入数据中最大的长度 |
+| `paddlenlp.data.Tuple`          | 将多个batchify函数包装在一起 |
+| `paddlenlp.data.SamplerHelper`  | 构建用于`Dataloader`的可迭代sampler |
+| `paddlenlp.data.Vocab`          | 用于文本token和ID之间的映射 |
+| `paddlenlp.data.JiebaTokenizer` | Jieba分词 |
+
+## API使用方法
+
+### `paddlenlp.data.Stack`
+
```
Comment (ZeyuChen) [inline PaddleNLP/docs/data.md:114]: 在这里补充下介绍更加好，API的说明

File: PaddleNLP/docs/data.md (hunk 1 above)
Comment (ZeyuChen) [reply PaddleNLP/docs/data.md:215]: 最后面再补充下与collate_fn结合的一些用法示例

**Review by ZeyuChen — CHANGES_REQUESTED** [review]
反馈了一些意见

File: PaddleNLP/docs/data.md (hunk 1 above)
Comment (xiemoyuan) [reply PaddleNLP/docs/data.md:215]: 已经修改了示例程序

File: PaddleNLP/docs/data.md (hunk 2 above)
Comment (xiemoyuan) [reply PaddleNLP/docs/data.md:114]: 已经修改了示例程序

File: PaddleNLP/docs/data.md [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+### `paddlenlp.data.JiebaTokenizer`
+
+`paddlenlp.data.JiebaTokenizer`初始化需传入`paddlenlp.data.Vocab`类，包含`cut`分词方法和将句子转换为ids的`encode`方法。
+
+```python
+from paddle.data import Vocab, JiebaTokenizer
+# 从文件构建词表
+vocab = Vocab.load_vocabulary(vocab_file_path)
+tokenizer = JiebaTokenizer(vocab)
+tokens = tokenizer.cut('我爱你中国')
```
Comment (ZeyuChen) [inline PaddleNLP/docs/data.md]: 像这种API，下面要加 
>>> 然后把对应结果print出来，可能更有API示例的感觉。

File: PaddleNLP/docs/data.md [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+### `paddlenlp.data.Pad`
+
+`paddlenlp.data.Pad`用来组建batch，它的输入长度不同，它首先会将输入数据全部padding到最大长度，然后再堆叠组成batch数据输出。
+
+```python
+from paddle.data import Pad
+a = [1, 2, 3, 4]
+b = [5, 6, 7]
+c = [8, 9]
+result = Pad(axis=0, pad_val=0)([a, b, c])
```
Comment (ZeyuChen) [inline PaddleNLP/docs/data.md]: axis的说明样例没体现出来

File: PaddleNLP/docs/data.md (hunk 1 above)
Comment (ZeyuChen) [reply PaddleNLP/docs/data.md:215]: 能有可视化的样例体现这个Tuple的作用吗？现在确实不太好理解

File: PaddleNLP/docs/data.md [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+
+`paddlenlp.data.SamplerHelper`的作用是构建用于`DataLoader`的可迭代采样器，它包含`shuffle`、`sort`、`batch`、`shard`等一系列方法，方便用户灵活使用。
+
+```python
+from paddle.data import SamplerHelper
+"""
+构建dataset
+dataset = ...
+"""
+batch_sampler = SamplerHelper(dataset).shuffle().batch(
```
Comment (ZeyuChen) [inline PaddleNLP/docs/data.md]: 需要示例，一个个清晰可感知的示例体现出
shuffle，sort，batch，shard的作用比较好

**Review by ZeyuChen — CHANGES_REQUESTED** [review]
@guoshengCS 进一步指导吧，SamplerHelper这块感觉还不够清晰

File: PaddleNLP/docs/data.md [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+    label = [label]
+    return ids, label
+
+dataset = GlueCoLA('train')
+dataset = MapDatasetWrapper(dataset).apply(convert_example, lazy=True)
+
+batchify_fn = lambda samples, fn=Tuple(
+    Pad(axis=0, pad_val=pad_id),  # ids
+    Stack(dtype='int64')  # label
+): fn(samples)
```
Comment (guoshengCS) [inline PaddleNLP/docs/data.md]: 这里应该直接用创建的Tuple实例就可以
```python
batchify_fn = Tuple(
    Pad(axis=0, pad_val=pad_id),  # ids
    Stack(dtype='int64')  # label
)
```

File: PaddleNLP/docs/data.md [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+
+```python
+from paddle.data import SamplerHelper
+"""
+构建dataset
+dataset = ...
+"""
+batch_sampler = SamplerHelper(dataset).shuffle().batch(
+    batch_size=batch_size,
+    drop_last=True)
```
Comment (guoshengCS) [inline PaddleNLP/docs/data.md]: 感觉drop_last用默认的False好些，一方面能稍微简短些，另一方面对特殊情况（不足batch size）的处理也能表现的更清楚些。

File: PaddleNLP/docs/data.md (hunk 5 above)
Comment (guoshengCS) [reply PaddleNLP/docs/data.md]: 这里可以用个简单的dataset，比如就是100到110的list，shuffle，sort，batch，shard的每一步后都写出迭代打印的结果，注意这里产生的是index的迭代器非数据本身

File: PaddleNLP/docs/data.md [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+bos_id = vocab.token_to_idx[vocab.bos_token]
+eos_id = vocab.token_to_idx[vocab.eos_token]
+
+def convert_example(example):
+    text, label = example
+    ids = [bos_id] + tokenizer.encode(text) + [eos_id]
+    label = [label]
+    return ids, label
+
+dataset = GlueCoLA('train')
```
Comment (guoshengCS) [inline PaddleNLP/docs/data.md]: 如果用JiebaTokenizer都话是否这里需要换用个中文数据集

File: PaddleNLP/docs/data.md [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+dataset = MapDatasetWrapper(dataset).apply(convert_example, lazy=True)
+
+batchify_fn = lambda samples, fn=Tuple(
+    Pad(axis=0, pad_val=pad_id),  # ids
+    Stack(dtype='int64')  # label
+): fn(samples)
+
+batch_size = 16
+batch_sampler = SamplerHelper(dataset).shuffle().batch(
+    batch_size=batch_size,
```
Comment (guoshengCS) [inline PaddleNLP/docs/data.md]: 这里就直接`batch_size=16`吧，示例文档能简洁些还是简洁些吧

File: PaddleNLP/docs/data.md [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+| API                             | 简介                                       |
+| ------------------------------- | :----------------------------------------- |
+| `paddlenlp.data.Stack`          | 堆叠N个具有相同shape的输入数据来构建一个batch |
+| `paddlenlp.data.Pad`            | 堆叠N个输入数据来构建一个batch，每个输入数据将会被padding到N个输入数据中最大的长度 |
+| `paddlenlp.data.Tuple`          | 将多个batchify函数包装在一起 |
+| `paddlenlp.data.SamplerHelper`  | 构建用于`Dataloader`的可迭代sampler |
+| `paddlenlp.data.Vocab`          | 用于文本token和ID之间的映射 |
+| `paddlenlp.data.JiebaTokenizer` | Jieba分词 |
+
+## API使用方法
```
Comment (guoshengCS) [inline PaddleNLP/docs/data.md:16]: 是否可以就这些API的作用分类说明，从dataloader中dataset、sampler和collate_fn三个参数展开。Stack/Pad/Tuple用来做collate_fn，SamplerHelper用来做batch_sampler，Tokenizer/Vocab用于dataset（对dataset做transform）。

File: PaddleNLP/docs/data.md [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+    def __getitem__(self, index):
+        data = self.data[index][0]
+        label = self.data[index][1]
+        return data, label
+
+    def __len__(self):
+        return len(self.data)
+
+dataset = MyDataset()
+# SamplerHelper返回的是数据的索引
```
Comment (guoshengCS) [inline PaddleNLP/docs/data.md]: `SamplerHelper返回的是数据索引的可迭代对象`  是否更准确些

File: PaddleNLP/docs/data.md [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+sampler = SamplerHelper(dataset) # [0, 1, 2]
+sampler = sampler.shuffle() # [0, 2, 1]
+key = (lambda x, data_source: len(data_source[x][0]))
+sampler = sampler.sort(key=key, buffer_size=2) # [2, 0, 1]
+sampler = sampler.batch(batch_size=2) #[[2, 0], [1]]
```
Comment (guoshengCS) [inline PaddleNLP/docs/data.md]: shuffle、sort、batch这里的注释说明可能还要更详细些
如sort可以说明是 `按照样本的首个字段的长度为key并在buffer_size大小个样本中排序，产生的迭代的索引为：[2, 0, 1]
`
shard的使用也可以在最后加上去

File: PaddleNLP/docs/data.md [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+
+```python
+from paddlenlp.data import Vocab
+# 从文件构建
+vocab1 = Vocab.load_vocabulary(vocab_file_path)
+# 从字典构建
+# dic = {'unk':0, 'pad':1, 'bos':2, 'eos':3, ...}
+vocab2 = Vocab.from_dict(dic)
+# 从json构建
+vocab3 = Vocab.from_json(json_str)
```
Comment (guoshengCS) [inline PaddleNLP/docs/data.md:42]: json的话会对能加载的格式有一定要求，配合to_json先to_json再from_json使用吧，

**Review by guoshengCS — APPROVED** [review]
